libsolidity/lsp/Utils.cpp

```diff
@@ -0,0 +1,70 @@
+#include <liblangutil/CharStreamProvider.h>
+#include <liblangutil/Exceptions.h>
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/lsp/FileRepository.h>
+#include <libsolidity/lsp/Utils.h>
+
+namespace solidity::lsp
+{
+
+using namespace frontend;
+using namespace langutil;
+using namespace std;
+
+optional<LineColumn> parseLineColumn(Json::Value const& _lineColumn)
+{
+	if (_lineColumn.isObject() && _lineColumn["line"].isInt() && _lineColumn["character"].isInt())
+		return LineColumn{_lineColumn["line"].asInt(), _lineColumn["character"].asInt()};
+	else
+		return nullopt;
+}
+
+Json::Value toJson(LineColumn _pos)
+{
+	Json::Value json = Json::objectValue;
+	json["line"] = max(_pos.line, 0);
+	json["character"] = max(_pos.column, 0);
+
+	return json;
+}
+
+Json::Value toJsonRange(LineColumn const& _start, LineColumn const& _end)
+{
+	Json::Value json;
+	json["start"] = toJson(_start);
+	json["end"] = toJson(_end);
+	return json;
+}
+
+vector<Declaration const*> allAnnotatedDeclarations(Expression const* _expression)
+{
+	vector<Declaration const*> output;
+
+	if (auto const* identifier = dynamic_cast<Identifier const*>(_expression))
+	{
+		output.push_back(identifier->annotation().referencedDeclaration);
+		output += identifier->annotation().candidateDeclarations;
+	}
+	else if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(_expression))
+	{
+		output.push_back(memberAccess->annotation().referencedDeclaration);
+	}
+
+	return output;
+}
+
+optional<SourceLocation> declarationPosition(Declaration const* _declaration)
```

Comment:
Why not `declarationLocation`?

Reply:
Also why do you return the name location? Is this how it's supposed to work?

Reply:
i've renamed the function.

With regards to `nameLocation`. If you want to goto a definition of a symbol, the editor/IDE will highlight the jumped-to area. If the whole location (i.e. with function body) is returned here, it'll all be highlighted. Which would be okay as it's only highlighted for a moment (in my editor), but the cursor is placed at the beginning, which would be the `f` of `function hello() ....` for example instead of the `h` of that `function hello() ...f`. Looking at other LSPs, it seems common to jump to the name of the definition, that's why we jump to the name location - if present - but fall back to the bigger-range location otherwise.